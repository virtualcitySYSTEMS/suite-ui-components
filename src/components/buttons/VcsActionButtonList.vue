<template>
  <div v-if="actions.length > 0">
    <VcsButton
      v-for="button in buttons"
      :key="button.name"
      :tooltip="button.title"
      :icon="button.icon"
      :active="button.active"
      @click="button.callback($event)"
      v-bind="{...$attrs}"
    />
    <v-menu
      v-if="overflowButtons.length > 0"
      content-class="vcs-overflow-menu-wrap"
      max-width="none"
      offset-x
    >
      <template #activator="{ on, attrs }">
        <VcsButton
          v-bind="{...$attrs, ...attrs}"
          v-on="on"
          class="vcs-overflow-btn"
        >
          <v-icon v-text="overflowIcon" />
        </VcsButton>
      </template>
      <VcsActionList :actions="overflowButtons" />
    </v-menu>
  </div>
</template>
<style lang="scss">
.vcs-overflow-menu-wrap{
  margin-left: 13px;
}
</style>
<script>

  import VcsButton from './VcsButton.vue';
  import VcsActionList, { validateActions } from '../lists/VcsActionList.vue';

  /**
   * @description
   * A component rendering a list of actions with overflow mechanic using
   * {@link VcsButton} and {@link VcsActionList}.
   * @vue-prop {Array<VcsAction>} actions - Array of actions
   * @vue-prop {number} [overflowCount=2] - number of buttons rendered until overflow.
   * @vue-prop {string} [overflowIcon='$vcsKebab'] - optional custom icon for overflow button
   * @vue-computed {Array<VcsAction>} buttons - buttons rendered directly, have to provide an icon
   * @vue-computed {Array<VcsAction>} overflowButtons - rest of buttons rendered in overflow
   */
  export default {
    name: 'VcsActionButtonList',
    components: { VcsActionList, VcsButton },
    props: {
      actions: {
        type: Array,
        required: true,
        validator: validateActions,
      },
      overflowCount: {
        type: Number,
        default: 2,
      },
      overflowIcon: {
        type: String,
        default: '$vcsKebab',
      },
    },
    computed: {
      buttons() {
        return this.actions.filter(i => i.icon).slice(0, this.overflowCount);
      },
      overflowButtons() {
        const buttonsNames = this.buttons.map(i => i.name);
        return this.actions.filter(i => !buttonsNames.includes(i.name));
      },
    },
  };
</script>
