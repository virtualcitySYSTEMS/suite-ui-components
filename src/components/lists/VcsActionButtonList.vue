<template>
  <div v-if="actions.length > 0">
    <VcsButton
      v-for="button in buttons"
      :key="button.name"
      :tooltip="button.title"
      :icon="button.icon"
      :active="button.active"
      :button-class="buttonClass"
      @click="button.callback($event)"
      class="pa-0"
    />
    <v-menu
      v-if="overflowButtons.length > 0"
      content-class="vcs-overflow-menu-wrap"
      max-width="none"
      offset-x
    >
      <template #activator="{ on, attrs }">
        <VcsButton
          v-bind="attrs"
          v-on="on"
          :button-class="buttonClass"
          class="vcs-overflow-btn"
        >
          <v-icon v-text="overflowIcon" />
        </VcsButton>
      </template>
      <v-list
        class="vcs-overflow-menu pa-0"
      >
        <v-list-item
          v-for="button in overflowButtons"
          :key="button.name"
          class="pa-0"
        >
          <VcsButton
            :tooltip="button.title"
            :icon="button.icon"
            :active="button.active"
            @click="button.callback($event)"
            button-class="list"
          >
            {{ button.name }}
          </VcsButton>
        </v-list-item>
      </v-list>
    </v-menu>
  </div>
</template>
<style lang="scss">

.vcs-overflow-menu-wrap{
  margin-left: 13px;
}
.vcs-overflow-menu{
  .v-list-item{
    min-height: 36px;
    border-bottom: 1px solid var(--v-accent-base);
  }
  .vcs-button-wrap{
    width: 100%;
  }
}
</style>
<script>
  import Vue from 'vue';
  import { is } from '@vcsuite/check';
  import VcsButton from '../buttons/VcsButton.vue';

  /**
   * @interface VcsAction
   * @property {string} name - reactive and translatable name rendered in overflow
   * @property {string} [title] - reactive and translatable title rendered as tooltip
   * @property {string} [icon] - icon rendered on the button. If no icon provided, item is rendered in overflow
   * @property {Function} callback - callback function is triggered when the button is clicked
   * @property {boolean} [active=false] - optional state of button. If active, button is rendered in primary color
   */

  /**
   * @description
   * A component rendering a list of action actions with overflow mechanic
   * @vue-prop {Array<VcsAction>} actions - Array of actions
   * @vue-prop {number} [overflowCount=2] - number of buttons rendered until overflow.
   * @vue-prop {string} [overflowIcon='$vcsKebab'] - optional custom icon for overflow button
   * @vue-prop {boolean} [iconShape=false] - whether to render button round as icon
   * @vue-prop {VcsButtonClass} [buttonClass=menu] - visual style of buttons
   * @vue-computed {Array<VcsAction>} buttons - buttons rendered directly, have to provide an icon
   * @vue-computed {Array<VcsAction>} overflowButtons - rest of buttons rendered in overflow
   */
  export default Vue.extend({
    name: 'VcsActionButtonList',
    components: { VcsButton },
    props: {
      actions: {
        type: Array,
        required: true,
        validator(actions) {
          return actions.every(item => is(item, {
            name: String,
            title: [undefined, String],
            icon: [undefined, String],
            callback: Function,
            active: [undefined, Boolean],
          }));
        },
      },
      overflowCount: {
        type: Number,
        default: 2,
      },
      overflowIcon: {
        type: String,
        default: '$vcsKebab',
      },
      iconShape: {
        type: Boolean,
        default: false,
      },
      buttonClass: {
        type: String,
        default: 'standard',
      },
    },
    computed: {
      buttons() {
        return this.actions.filter(i => i.icon).slice(0, this.overflowCount);
      },
      overflowButtons() {
        const buttonsNames = this.buttons.map(i => i.name);
        return this.actions.filter(i => !buttonsNames.includes(i.name));
      },
    },
  });
</script>
