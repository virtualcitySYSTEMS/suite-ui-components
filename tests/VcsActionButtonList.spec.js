import { createLocalVue, mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import Vuetify from 'vuetify';
import VcsActionButtonList from '../src/components/buttons/VcsActionButtonList.vue';

describe('VcsActionButtonList', () => {
  let localVue;
  let vuetify;
  let actions;
  let cmp;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    actions = [
      { name: 'foo', callback() {} },
      { name: 'bar', callback() {} },
      { name: 'icon', icon: '$vcsIcon', callback() {} },
      { name: 'icon2', icon: '$vcsIcon2', callback() {} },
    ];
    cmp = mount(VcsActionButtonList, {
      localVue,
      vuetify,
      propsData: {
        actions,
      },
    });
  });

  afterEach(() => cmp.destroy());

  it('should mount', () => {
    expect(cmp.html()).toMatchSnapshot();
  });

  describe('compute buttons and overflow buttons', () => {
    it('should filter actions with no icon', () => {
      expect(cmp.vm.actions).toMatchObject(actions);
      expect(cmp.vm.buttons).toMatchObject([actions[2], actions[3]]);
      expect(cmp.vm.overflowButtons).toMatchObject([actions[0], actions[1]]);
    });

    it('should filter by overflowCount', () => {
      cmp.vm.$props.overflowCount = 1;
      expect(cmp.vm.actions).toMatchObject(actions);
      expect(cmp.vm.buttons).toMatchObject([actions[2]]);
      expect(cmp.vm.overflowButtons).toMatchObject([actions[0], actions[1], actions[3]]);
    });
  });
});
