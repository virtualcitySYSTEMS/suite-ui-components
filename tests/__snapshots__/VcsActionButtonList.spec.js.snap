// Vitest Snapshot v1

exports[`VcsActionButtonList > should mount 1`] = `
"<div>
  <div data-v-4027582a=\\"\\" class=\\"d-inline-block vcs-button-wrap pa-0\\"><button data-v-4027582a=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--text theme--light v-size--default vcs-button--standard padding font-weight-bold\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\"><span class=\\"v-btn__content\\"><i data-v-4027582a=\\"\\" aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">$vcsIcon</i></span></button><span data-v-62384dbb=\\"\\" data-v-4027582a=\\"\\" class=\\"v-tooltip v-tooltip--bottom\\" tooltipposition=\\"bottom\\"><!----></span>
    <!---->
  </div>
  <div data-v-4027582a=\\"\\" class=\\"d-inline-block vcs-button-wrap pa-0\\"><button data-v-4027582a=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--text theme--light v-size--default vcs-button--standard padding font-weight-bold\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\"><span class=\\"v-btn__content\\"><i data-v-4027582a=\\"\\" aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">$vcsIcon2</i></span></button><span data-v-62384dbb=\\"\\" data-v-4027582a=\\"\\" class=\\"v-tooltip v-tooltip--bottom\\" tooltipposition=\\"bottom\\"><!----></span>
    <!---->
  </div>
  <div class=\\"v-menu\\">
    <!---->
  </div>
  <div data-v-4027582a=\\"\\" class=\\"d-inline-block vcs-button-wrap vcs-overflow-btn\\"><button data-v-4027582a=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--text theme--light v-size--default vcs-button--standard padding font-weight-bold\\" role=\\"button\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\"><span class=\\"v-btn__content\\"><!----><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">$vcsKebab</i></span></button><span data-v-62384dbb=\\"\\" data-v-4027582a=\\"\\" class=\\"v-tooltip v-tooltip--bottom\\" tooltipposition=\\"bottom\\"><!----></span>
    <!---->
  </div>
</div>"
`;
